bee-network/bee-autopeering/src/peering/filter.rs

```diff
@@ -0,0 +1,151 @@
+
+use crate::{
+    peer::{peer_id::PeerId, Peer},
+    NeighborValidator,
+};
+
+use std::{
+    collections::HashSet,
+    sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard},
+};
+
+#[derive(Clone)]
+pub(crate) struct NeighborFilter<V: NeighborValidator> {
+    inner: Arc<RwLock<NeighborFilterInner<V>>>,
+}
+
+impl<V: NeighborValidator> NeighborFilter<V> {
+    /// Creates a new filter.
+    ///
+    /// A peer id same as `local_id` will always be rejected. A `validator` must be provided
+    /// to inject a user defined criteria.
+    pub fn new(local_id: PeerId, validator: V) -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(NeighborFilterInner::new(local_id, validator))),
+        }
+    }
+
+    /// Adds a single peer id that should be rejected.
+    pub(crate) fn add(&self, peer_id: PeerId) {
+        self.write().add(peer_id);
+    }
+
+    /// Resets the filter (i.e. removes all currently rejected peer ids).
+    pub(crate) fn clear(&self) {
+        self.write().clear();
+    }
+
+    /// Returns `true` if the filter is okay with the candidate, otherwise `false`.
+    pub(crate) fn ok(&self, candidate: impl AsRef<Peer>) -> bool {
+        self.read().ok(candidate)
+    }
+
+    /// Applies the filter to a list of candidates.
+    pub(crate) fn apply_list<'a, P: AsRef<Peer>>(&self, candidates: &'a [P]) -> Vec<&'a P> {
+        self.read().apply_list(candidates)
+    }
+
+    fn read(&self) -> RwLockReadGuard<NeighborFilterInner<V>> {
+        self.inner.read().expect("error getting read access")
+    }
+
+    fn write(&self) -> RwLockWriteGuard<NeighborFilterInner<V>> {
+        self.inner.write().expect("error getting write access")
+    }
+}
+
+pub(crate) struct NeighborFilterInner<V: NeighborValidator> {
+    local_id: PeerId,
+    rejected: HashSet<PeerId>,
+    validator: V,
+}
+
+impl<V: NeighborValidator> NeighborFilterInner<V> {
+    fn new(local_id: PeerId, validator: V) -> Self {
+        Self {
+            local_id,
+            rejected: HashSet::new(),
+            validator,
+        }
+    }
+
+    /// Adds a single peer id that should be rejected.
+    fn add(&mut self, peer_id: PeerId) {
+        self.rejected.insert(peer_id);
+    }
+
+    /// Resets the filter (i.e. removes all currently rejected peer ids).
+    fn clear(&mut self) {
+        self.rejected.clear()
+    }
+
+    /// Returns `true` if the filter is okay with the candidate, otherwise `false`.
+    fn ok(&self, candidate: impl AsRef<Peer>) -> bool {
+        let peer = candidate.as_ref();
+        let peer_id = peer.peer_id();
+
+        if peer_id == &self.local_id || self.rejected.contains(peer_id) {
+            false
+        } else {
+            self.validator.is_valid(peer)
+        }
+    }
+
+    /// Applies the filter to a list of candidates.
+    fn apply_list<'a, P: AsRef<Peer>>(&self, candidates: &'a [P]) -> Vec<&'a P> {
+        candidates.iter().filter(|c| self.ok(*c)).collect::<Vec<_>>()
```

Comment:
Is this `*c` necessary ? Can't figure out the types

Reply:
It's double-ref, and Rust can figure it out automatically, so it's not necessary.
Removed.